### PR TITLE
CI: Fix IPython Tab Completion test async warning

### LINKED ISSUE #29070: Warnings from IPython tab completion tests
At some point, IPython changed their code to be async. Now, our tests using their tab completion stuff causes warnings.

```
$ pytest pandas/tests/ -k test_tab_complete_warning 

================================================================================= warnings summary =================================================================================
pandas/tests/arrays/categorical/test_warnings.py::TestCategoricalWarnings::test_tab_complete_warning
  /Users/taugspurger/sandbox/pandas/pandas/tests/arrays/categorical/test_warnings.py:14: RuntimeWarning: coroutine 'InteractiveShell.run_code' was never awaited
    ip.run_code(code)

pandas/tests/frame/test_api.py::TestDataFrameMisc::test_tab_complete_warning
  /Users/taugspurger/sandbox/pandas/pandas/tests/frame/test_api.py:575: RuntimeWarning: coroutine 'InteractiveShell.run_code' was never awaited
    ip.run_code(code)

pandas/tests/indexes/test_base.py::TestIndex::test_tab_complete_warning
  /Users/taugspurger/sandbox/pandas/pandas/tests/indexes/test_base.py:2420: RuntimeWarning: coroutine 'InteractiveShell.run_code' was never awaited
    ip.run_code(code)

pandas/tests/series/test_api.py::TestSeriesMisc::test_tab_complete_warning
  /Users/taugspurger/sandbox/pandas/pandas/tests/series/test_api.py:508: RuntimeWarning: coroutine 'InteractiveShell.run_code' was never awaited
    ip.run_code(code)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=========================================================== 4 passed, 11 skipped, 62860 deselected, 4 warnings in 29.50s ===========================================================

```

There's a race condition between the completion of the `ip.run_code(code)` calls and the asserts later on. We need to explicitly wait for the `ip.run_code(code)` to finish before moving on with the test. The easiest way is to

1. Mark the tests as async with the 3rd-party pytest-asyncio (need to add to the CI)
2. Rewrite the tests to use `async def`.
3. Add an `await` to the `ip.run_code`

```python
diff --git a/pandas/tests/series/test_api.py b/pandas/tests/series/test_api.py
index 998f8b6f7d..b471e1dc1e 100644
--- a/pandas/tests/series/test_api.py
+++ b/pandas/tests/series/test_api.py
@@ -499,13 +499,15 @@ class TestSeriesMisc(TestData, SharedWithSparse):
         for full_series in [pd.Series([1]), pd.Series(index=[1])]:
             assert not full_series.empty
 
-    def test_tab_complete_warning(self, ip):
+    @pytest.mark.asyncio
+    async def test_tab_complete_warning(self, ip):
         # https://github.com/pandas-dev/pandas/issues/16409
         pytest.importorskip("IPython", minversion="6.0.0")
         from IPython.core.completer import provisionalcompleter
 
         code = "import pandas as pd; s = pd.Series()"
-        ip.run_code(code)
+        await ip.run_code(code)
+
         with tm.assert_produces_warning(None):
             with provisionalcompleter("ignore"):
                 list(ip.Completer.completions("s.", 1))
```

~This is currently blocked by https://github.com/pandas-dev/pandas/issues/29034. Once we drop 3.5, we can use the `async` and `await` syntax.~ I forgot 3.5 had async / await.

### PR DESCRIPTION
**Pull request by alimcmaster1**
- [x] closes #29070
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

- New PR since https://github.com/pandas-dev/pandas/pull/29087 went stale

### CODE REVIEW

File: pandas/conftest.py [hunk 1]
```diff
@@ -930,3 +931,13 @@ def __len__(self):
             return self._data.__len__()
 
     return TestNonDictMapping
+
+
+def async_mark():
+    try:
+        import_optional_dependency("pytest_asyncio")
```
Comment (jreback) [inline pandas/conftest.py]: don't put this here, rather in _test_decorators

File: pandas/tests/arrays/categorical/test_warnings.py [hunk 2]
```diff
@@ -1,16 +1,18 @@
 import pytest
 
 import pandas._testing as tm
+from pandas.conftest import async_mark
```
Comment (jreback) [inline pandas/tests/arrays/categorical/test_warnings.py]: see comments above, we don't want to import from pandas.conftest at all.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/conftest.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 import pytest
 from pytz import FixedOffset, utc
 
+from pandas.compat._optional import import_optional_dependency
```
Comment (jreback) [inline pandas/conftest.py]: don't need this import any longer

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/conftest.py (hunk 3 above)
Comment (alimcmaster1) [reply pandas/conftest.py]: yep sorry just ran flake8 ...

File: pandas/util/_test_decorators.py [hunk 4]
```diff
@@ -251,3 +252,13 @@ def new_func(*args, **kwargs):
         assert flist2 == flist
 
     return new_func
+
+
+def async_mark():
```
Comment (jreback) [inline pandas/util/_test_decorators.py]: in a future PR can you type & add a doc-string

**Review by jreback — APPROVED** [review]
lgtm.

**Comment by jreback** [comment]
thanks @alimcmaster1

**Pull request merged**